versions/draft-2/parsers/antlr4/WdlParser.g4

```diff
@@ -0,0 +1,272 @@
+	;
+
+expr_infix3
+	: expr_infix3 PLUS expr_infix4 #add
+	| expr_infix3 MINUS expr_infix4 #sub
+	| expr_infix4 #infix4
+	;
+
+expr_infix4
+	: expr_infix4 STAR expr_infix5 #mul
+	| expr_infix4 DIVIDE expr_infix5 #divide
+	| expr_infix4 MOD expr_infix5 #mod
+	| expr_infix5 #infix5
+	;
+
+expr_infix5
+	: expr_core
+	;
+
+expr_core
+	: LPAREN expr RPAREN #expression_group
+	| primitive_literal #primitives
+	| LBRACK (expr (COMMA expr)*)* RBRACK #array_literal
+	| LPAREN expr COMMA expr RPAREN #pair_literal
+	| LBRACE (expr COLON expr (COMMA expr COLON expr)*)* RBRACE #map_literal
+	| OBJECT_LITERAL LBRACE (Identifier COLON expr (COMMA Identifier COLON expr)*)* RBRACE #object_literal
+	| NOT expr #negate
+	| (PLUS | MINUS) expr #unirarysigned
+	| expr_core LBRACK expr RBRACK #at
+	| IF expr THEN expr ELSE expr #ifthenelse
+	| Identifier LPAREN (expr (COMMA expr)*)? RPAREN #apply
+	| Identifier #left_name
+	| expr_core DOT Identifier #get_name
+	;
+
+import_alias
+	: ALIAS Identifier AS Identifier
+	;
+
+import_as
+	: AS Identifier
+	;
+
+import_doc
+	: IMPORT string import_as? (import_alias)*
+	;
+
+meta_kv
+	: Identifier COLON string
+	;
+
+parameter_meta
+	: PARAMETERMETA LBRACE meta_kv* RBRACE
+	;
+
+meta
+	:  META LBRACE meta_kv* RBRACE
+	;
+
+task_runtime_kv
+	: Identifier COLON expr
+	;
+
+task_runtime
+	: RUNTIME LBRACE (task_runtime_kv)* RBRACE
+	;
+
+task_input
+	: (any_decls)*
+	;
+
+task_output
+	: OUTPUT LBRACE (bound_decls)* RBRACE
+	;
+
+task_command_string_part
+	: CommandStringPart*
+	;
+
+task_command_expr_part
+	: StringCommandStart  (expression_placeholder_option)* expr RBRACE
+	;
+
+task_command_expr_with_string
+	: task_command_expr_part task_command_string_part
+	;
+
+task_command
+	: COMMAND task_command_string_part task_command_expr_with_string* EndCommand
+	| HEREDOC_COMMAND task_command_string_part task_command_expr_with_string* EndCommand
+	;
+
+task_element
+	: task_output
+	| task_command
+	| task_runtime
+	| parameter_meta
+	| meta
+	| any_decls
+	;
+
+task
+	: TASK Identifier LBRACE task_input (task_element)+ RBRACE
+	;
+
+inner_workflow_element
+	: bound_decls
+	| call
+	| scatter
+	| conditional
+	;
+
+call_alias
+	: AS Identifier
+	;
+
+call_input
+	: Identifier EQUAL expr
+	;
+
+call_inputs
+	: INPUT COLON (call_input (COMMA call_input)*)
+	;
+
+call_body
+	: LBRACE call_inputs? RBRACE
+	;
+
+call_name
+	: Identifier (DOT Identifier)*
+	;
+
+call
+	: CALL call_name call_alias?  call_body?
+	;
+
+scatter
+	: SCATTER LPAREN Identifier In expr RPAREN LBRACE inner_workflow_element* RBRACE
+	;
+
+conditional
+	: IF LPAREN expr RPAREN LBRACE inner_workflow_element* RBRACE
+	;
+
+workflow_input
+	: (any_decls)*
+	;
+
+workflow_output
+	: OUTPUT LBRACE (bound_decls)* RBRACE
+	;
+
+workflow_element
+	: workflow_output #output
+	| inner_workflow_element #inner_element
+	| parameter_meta #parameter_meta_element
+	| meta #meta_element
+	;
+
+workflow
+	: WORKFLOW Identifier LBRACE workflow_input workflow_element* RBRACE
```

Comment:
Similar situation to the task above. Workflow inputs are not bound to be defined at the start of the workflow, but can be interspersed throughout as either a declaration or a bound declartion.  moving the `workflow_input` to the workflow_element and getting rid of the `bound_decls` from the `workflow_element` rule will fix this

Reply:
"Tasks define all their inputs as declarations at the top of the task definition." https://github.com/openwdl/wdl/blob/master/versions/draft-2/SPEC.md#task-inputs

IMO this should be interpreted literally - any declaration before the first non-declaration is considered an input, and others must be bound and are added to bound_decls.

Reply:
hmm good point. I guess this is one of the ambiguities that later versions hoped to resolve with the `input {...}` block.